server/util.go

```diff
@@ -276,31 +287,110 @@ func (rf *redirectFormatter) Format(pkg string, level capnslog.LogLevel, depth i
 // Flush only for implementing Formatter.
 func (rf *redirectFormatter) Flush() {}
 
+// modifyHook injects file name and line pos into log entry.
+type contextHook struct{}
+
+// Fire implements logrus.Hook interface
+// https://github.com/sirupsen/logrus/issues/63
+func (hook *contextHook) Fire(entry *log.Entry) error {
+	pc := make([]uintptr, 3, 3)
+	cnt := runtime.Callers(6, pc)
+
+	for i := 0; i < cnt; i++ {
+		fu := runtime.FuncForPC(pc[i] - 1)
+		name := fu.Name()
+		if !strings.Contains(name, "github.com/Sirupsen/logrus") &&
+			!strings.Contains(name, "github.com/coreos/pkg/capnslog") {
+			file, line := fu.FileLine(pc[i] - 1)
+			entry.Data["file"] = path.Base(file)
+			entry.Data["line"] = line
+			break
+		}
+	}
+	return nil
+}
+
+// Levels implements logrus.Hook interface.
+func (hook *contextHook) Levels() []log.Level {
+	return log.AllLevels
+}
+
+func stringToLogLevel(level string) log.Level {
+	switch strings.ToLower(level) {
+	case "fatal":
+		return log.FatalLevel
+	case "error":
+		return log.ErrorLevel
+	case "warn":
+		return log.WarnLevel
+	case "warning":
+		return log.WarnLevel
+	case "debug":
+		return log.DebugLevel
+	case "info":
+		return log.InfoLevel
+	}
+	return defaultLogLevel
+}
+
+// for compatability with ngaut/log
+type textFormatter struct{}
+
+// Format implements logrus.Formatter
+func (f *textFormatter) Format(entry *log.Entry) ([]byte, error) {
+	var b *bytes.Buffer
+	if entry.Buffer != nil {
+		b = entry.Buffer
+	} else {
+		b = &bytes.Buffer{}
+	}
+	b.WriteString(entry.Time.Format("2006/01/02 15:04:05"))
+	if file, ok := entry.Data["file"]; ok {
+		b.WriteString(fmt.Sprintf(" %s:%v:", file, entry.Data["line"]))
```

Comment:
seem you can use fmt.Fprintf(b, format, args...) directly to reduce the allocation cost for fmt.Sprintf?